{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "colab_type": "text",
        "id": "view-in-github"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/jasonwong-lab/HKU-Practical-Bioinformatics/blob/main/NGS_sequence_alignment_command_line.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "o_lO85QksbvE"
      },
      "source": [
        "### BBMS3004 2026 - Targeted sequencing in acute myeloid leukaemia\n",
        "\n",
        "*by Alvin Ip*\n",
        "\n",
        "This practical aims to demonstrate how next-generation sequencing data is processed to yield clinically useful information. It is a highly condensed version of BIOF3002 Clinical Genomics and Bioinformatics that Prof Jason Wong and I co-coordinated for bioinformatics students.\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "7gnJkcA9tTBn"
      },
      "source": [
        "## *** Package installation and downloads for workshop (~ 10 minutes)\n",
        "\n",
        "1.   conda (for simple installation of packages)\n",
        "2.   FastQC (for reads quality check)\n",
        "3.   Trimmomatic (for trimming low-quality bases and adapters)\n",
        "4.   bwa (tools for sequence alignment)\n",
        "5.   samtools (tools for processing sam & bam files)  \n",
        "\n",
        "**IMPORTANT：Every time you connect to Google Colab, you have to perform these set up steps again.**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Install terminal for Google Colab\n",
        "!pip install colab-xterm\n",
        "%load_ext colabxterm\n",
        "%xterm"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "AOrw7MhPztUW"
      },
      "outputs": [],
      "source": [
        "# Set working pathway to your own Google drive, so that the data files are saved after closing Colab (~ 1 min)\n",
        "from google.colab import drive\n",
        "drive.mount('/content/gdrive')"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "B_5lJh21K7Ya"
      },
      "outputs": [],
      "source": [
        "# Install igv-notebook (<1 min)\n",
        "!pip install igv-notebook"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "UJMUyjxntquy"
      },
      "outputs": [],
      "source": [
        "# Install conda (~ 1 min). There will be a message saying that the session has crashed, but don't worry about this. This is due to the session restarting following conda installation.\n",
        "!pip install -q condacolab\n",
        "import condacolab\n",
        "condacolab.install()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "F5J8UCIn7i7H"
      },
      "outputs": [],
      "source": [
        "# Install fastqc (~ 2 mins)\n",
        "!conda install -c bioconda fastqc"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "hkiiT0YRz4S2"
      },
      "outputs": [],
      "source": [
        "# Install bwa (~ 1 min)\n",
        "!conda install -c bioconda bwa"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "oMH5G73bz_CI"
      },
      "outputs": [],
      "source": [
        "# Install samtools (~1 min)\n",
        "#!conda install -c bioconda samtools"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Install samtools, bcftools, htslib, bedtools, bamtools (~ 2 mins)\n",
        "!conda install -y -c conda-forge -c bioconda \\\n",
        "  samtools=1.20 bcftools=1.20 htslib=1.20 bedtools=2.31.1 bamtools"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Check installation ran correctly\n",
        "!samtools --version | head -1"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Install GATK in a new environment (to specify the exapct openjdk and python version) ~1.5 mins\n",
        "!conda create -y -n gatkenv python=3.10\n",
        "!conda config --env --set channel_priority strict\n",
        "!conda install -y -n gatkenv -c conda-forge -c bioconda gatk4=4.6.2.0 openjdk=17"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Check that gatk is installed properly\n",
        "!conda run -n gatkenv gatk --version"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "#Install Trimmomatic\n",
        "#!apt-get install openjdk-8-jre-headless -qq  # Install Java\n",
        "#!wget http://www.usadellab.org/cms/uploads/supplementary/Trimmomatic/Trimmomatic-0.39.zip  # Download Trimmomatic\n",
        "#!unzip Trimmomatic-0.39.zip  # Unzip the downloaded file"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "jQ4sstuDsBO1"
      },
      "source": [
        "## Set working directory\n",
        "\n",
        "By default, the working directory will be My Drive/bbms3004"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "p4_KlYtbrzzF"
      },
      "outputs": [],
      "source": [
        "import os\n",
        "try:\n",
        "  os.mkdir(\"/content/gdrive/My Drive/bbms3004\")         # change this path if necessary\n",
        "except FileExistsError:\n",
        "  print(\"directory already exist. OK to continue\")\n",
        "os.chdir(\"/content/gdrive/My Drive/bbms3004\")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "TPs7Tb-tyxc_"
      },
      "source": [
        "## Download ready prepared files for analysis."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "collapsed": true,
        "id": "OwoP1gu3GWLX"
      },
      "outputs": [],
      "source": [
        "# Download data files from Google Drive if not already present (~ 3 min)\n",
        "# Double check that we are in the right directory (~ 30s)\n",
        "import os\n",
        "os.chdir(\"/content/gdrive/MyDrive/bbms3004\")                     # change this path if necessary\n",
        "\n",
        "import os\n",
        "if os.path.isfile(\"/content/gdrive/MyDrive/bbms3004/bbms3004_toolkit/17H0510082_1.fastq.gz\"):    # check if the file exist\n",
        "  print(\"Data files already exist, OK to continue.\")\n",
        "else:\n",
        "  !pip install gdown\n",
        "  #!gdown -O bbms3004_toolkit.zip https://drive.google.com/uc?id=1aRJVznjy5WLQ5Dc0DT9c6NiXw64HdoKr # download if file not exist\n",
        "  !gdown -O bbms3004_toolkit.zip https://drive.google.com/uc?id=19Ql6GbWow2jj77UvQwCBMevnOmRs6S4I # download if file not exist\n",
        "  # unzip toolkit files\n",
        "  !unzip -o bbms3004_toolkit.zip\n",
        "  # remove the downloaded zip file after extraction\n",
        "  !rm bbms3004_toolkit.zip\n",
        "\n",
        "!ls -l ./bbms3004_toolkit/"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "XB9TrCTC7Ncy"
      },
      "source": [
        "## NGS_alignment command line\n",
        "\n",
        "1.1 Quality control\n",
        "\n",
        "1.2. Sequence alignment using Burrows–Wheeler Aligner (BWA)\n",
        "\n",
        "1.3. Viewing SAM files\n",
        "\n",
        "1.4. Align the WGS paired-end file"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "4dYhhxiG4lMM"
      },
      "outputs": [],
      "source": [
        "# Check what datasets we have downloaded\n",
        "%cd /content/gdrive/MyDrive/bbms3004/bbms3004_toolkit\n",
        "!ls -l\n",
        "\n",
        "%cd data\n",
        "!ls -l"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "EMrO9wOVnNH2"
      },
      "outputs": [],
      "source": [
        "# To look at the ChIP-seq fastq file type (head -n 12 to print first 12 lines):\n",
        "!zcat 17H0510082_1.fastq.gz | head -n 12"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "t4g-To8mPw8V"
      },
      "outputs": [],
      "source": [
        "# How many reads are there in the file? Type your code below:\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "pxBspjPp7Dms"
      },
      "source": [
        "### Quality control"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "f82vAykr7Fie"
      },
      "outputs": [],
      "source": [
        "## Run FastQC\n",
        "!fastqc 17H0510082_1.fastq.gz\n",
        "!fastqc 17H0510082_2.fastq.gz"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "7v4bH04sZtn5"
      },
      "outputs": [],
      "source": [
        "# Check the html file output by FastQC\n",
        "!ls\n",
        "\n",
        "# Download the html file and check it on your local browser\n",
        "from google.colab import files\n",
        "files.download('17H0510082_1_fastqc.html')"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Gpt-qEVO81Bs"
      },
      "source": [
        "You can also download the file **17H0510082_1_fastqc.html** from \"gdrive/MyDrive/bbms3004/Datasets\" on the left side."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Perform Trimmomatic\n",
        "!java -jar ../Trimmomatic-0.39/trimmomatic-0.39.jar PE -phred33 \\\n",
        "    17H0510082_1.fastq.gz 17H0510082_2.fastq.gz \\\n",
        "    17H0510082_1_trimmed_paired.fq.gz 17H0510082_1_trimmed_unpaired.fq.gz 17H0510082_2_trimmed_paired.fq.gz 17H0510082_2_trimmed_unpaired.fq.gz \\\n",
        "    ILLUMINACLIP:../Trimmomatic-0.39/adapters/TruSeq3-PE-2.fa:2:30:10 LEADING:10 TRAILING:10 SLIDINGWINDOW:4:15 MINLEN:40\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "mwdi1hJ59Dtu"
      },
      "source": [
        "### Burrows–Wheeler Aligner"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "yE-r2cjw8NfS"
      },
      "outputs": [],
      "source": [
        "# Look at the options for bwa and bwa mem\n",
        "!bwa"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "EQCq422B9HbZ"
      },
      "outputs": [],
      "source": [
        "!bwa mem"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "AuteGS6HdjLs"
      },
      "outputs": [],
      "source": [
        "# Let's take a look at the database files\n",
        "%cd /content/gdrive/My Drive/bbms3004/bbms3004_toolkit/Ref\n",
        "\n",
        "!ls -l"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "aFb8h3SZNyjY"
      },
      "outputs": [],
      "source": [
        "# Get ready to run BWA: First go into the Datasets directory\n",
        "%cd /content/gdrive/My Drive/bbms3004/bbms3004_toolkit/data"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "8EBR94fLMsQI"
      },
      "outputs": [],
      "source": [
        "# Do sequence alignment with the default options\n",
        "#!bwa mem ../bbms3004_toolkit/chr2.fa ./ChIP-seq_H3K27ac_example.fq.gz > ./BAM/ChIP-seq_H3K27ac_example.sam\n",
        "!bwa mem -t 4 -M ../Ref/ucsc.hg19.subset.fasta 17H0510082_1_trimmed_paired.fq.gz 17H0510082_2_trimmed_paired.fq.gz > 17H0510082.sam"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "8oqpLEwbNsCd"
      },
      "outputs": [],
      "source": [
        "# Check the result\n",
        "!head -n 20 17H0510082.sam"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "BEQNtRCyOSOk"
      },
      "source": [
        "### Working with SAM files"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "KLBxHt5COTho"
      },
      "outputs": [],
      "source": [
        "# Check the samtools command\n",
        "!samtools"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Sort the SAM file and compress SAM to BAM\n",
        "!conda run -n gatkenv gatk SortSam \\\n",
        "    -I 17H0510082.sam \\\n",
        "    -O 17H0510082_sorted.bam \\\n",
        "    --SORT_ORDER coordinate"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Add read groups (RG), tags for downstream tools to run properly\n",
        "!conda run -n gatkenv gatk AddOrReplaceReadGroups \\\n",
        "    -I 17H0510082_sorted.bam \\\n",
        "    -O 17H0510082_RG.bam \\\n",
        "    --RGID SPACE --RGLB panel --RGPL ILLUMINA --RGPU unit1 --RGSM 17H0510082 --CREATE_INDEX true"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Perform MarkDuplicates to mark duplicate reads\n",
        "!conda run -n gatkenv gatk MarkDuplicates \\\n",
        "    -I 17H0510082_RG.bam \\\n",
        "    -O 17H0510082_MD.bam \\\n",
        "    --METRICS_FILE 17H0510082_MD.stats \\\n",
        "    --CREATE_INDEX true"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Perform BaseRecalibrator to recalibrate base quality scores\n",
        "!conda run -n gatkenv gatk BaseRecalibrator \\\n",
        "    -I 17H0510082_MD.bam \\\n",
        "    -R ../Ref/ucsc.hg19.subset.fasta \\\n",
        "    --known-sites../Ref/dbsnp_138.hg19.vcf \\\n",
        "    --known-sites ../Ref/Mills_and_1000G_gold_standard.indels.hg19.vcf \\\n",
        "    -L ../Ref/myeloid-targets.interval_list \\\n",
        "    -ip 50 \\\n",
        "    -O 17H0510082_recal_data.table\n",
        "\n",
        "# Apply the recalibration to the sequence data\n",
        "!conda run -n gatkenv gatk ApplyBQSR \\\n",
        "    -R ../Ref/ucsc.hg19.subset.fasta \\\n",
        "    -I 17H0510082_MD.bam \\\n",
        "    --bqsr-recal-file 17H0510082_recal_data.table \\\n",
        "    -O 17H0510082_BR.bam"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Collect QC metrics\n",
        "!conda run -n gatkenv gatk CollectMultipleMetrics \\\n",
        "    -I 17H0510082_BR.bam \\\n",
        "    -R ../Ref/ucsc.hg19.subset.fasta \\\n",
        "    -O 17H0510082_QC_metrics\n",
        "\n",
        "# Collect read counts\n",
        "!conda run -n gatkenv gatk CollectReadCounts \\\n",
        "    -I 17H0510082_BR.bam \\\n",
        "    -L ../Ref/myeloid-targets.interval_list \\\n",
        "    --interval-merging-rule OVERLAPPING_ONLY \\\n",
        "    --format TSV \\\n",
        "    -O 17H0510082_read_counts.table\n",
        "\n",
        "# Collect HS metrics\n",
        "!conda run -n gatkenv gatk CollectHsMetrics \\\n",
        "    -I 17H0510082_BR.bam \\\n",
        "    -O 17H0510082_hs_metrics.txt \\\n",
        "    -R ../Ref/ucsc.hg19.subset.fasta \\\n",
        "    -BAIT_INTERVALS ../Ref/myeloid-probe-coords.interval_list \\\n",
        "    -TARGET_INTERVALS ../Ref/myeloid-targets.interval_list"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Perform variant calling with Mutect2\n",
        "!conda run -n gatkenv gatk Mutect2 \\\n",
        "    -R ../Ref/ucsc.hg19.subset.fasta \\\n",
        "    -I 17H0510082_BR.bam \\\n",
        "    -L ../Ref/myeloid-targets.interval_list \\\n",
        "    --germline-resource ../Ref/af-only-gnomad.myeloid.bedtools.vcf.gz \\\n",
        "    --f1r2-tar-gz f1r2.tar.gz \\\n",
        "    -O 17H0510082_unfiltered.vcf.gz\n",
        "\n",
        "!conda run -n gatkenv gatk LearnReadOrientationModel \\\n",
        "    -I f1r2.tar.gz \\\n",
        "    -O read-orientation-model.tar.gz\n",
        "\n",
        "!conda run -n gatkenv gatk GetPileupSummaries \\\n",
        "    -I 17H0510082_BR.bam \\\n",
        "    -V ../Ref/small_exac_common_myeloid.vcf.gz \\\n",
        "    -L ../Ref/small_exac_common_myeloid.vcf.gz  \\\n",
        "    -O getpileupsummaries.table\n",
        "\n",
        "!conda run -n gatkenv gatk CalculateContamination \\\n",
        "    -I getpileupsummaries.table \\\n",
        "    -O contamination.table\n",
        "\n",
        "!conda run -n gatkenv gatk FilterMutectCalls \\\n",
        "    -V 17H0510082_unfiltered.vcf.gz \\\n",
        "    -R ../Ref/ucsc.hg19.subset.fasta \\\n",
        "    --contamination-table contamination.table \\\n",
        "    --ob-priors read-orientation-model.tar.gz \\\n",
        "    -O 17H0510082_filtered.vcf.gz"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Use bcftools to filter variants with variant allele frequencies less than 0.05\n",
        "!bcftools filter -i \" FORMAT/AF \" 17H0510082_filtered.vcf.gz -o 17H0510082_filtered_AF05.vcf"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Perform variant annotation by ANNOVAR\n",
        "!perl ../annovar/table_annovar.pl 17H0510082_filtered_AF05.vcf \\\n",
        "    -buildver hg19 \\\n",
        "    -out 17H0510082_filtered_annotate \\\n",
        "    -remove \\\n",
        "    -protocol refGene,cosmic86,clinvar_20170905,exac03nontcga,gnomad_exome \\\n",
        "    -operation g,f,f,f,f \\\n",
        "    -nastring . \\\n",
        "    -vcfinput"
      ]
    }
  ],
  "metadata": {
    "colab": {
      "include_colab_link": true,
      "provenance": []
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}
