{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "colab_type": "text",
        "id": "view-in-github"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/jasonwong-lab/HKU-Practical-Bioinformatics/blob/main/NGS_sequence_alignment_command_line.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "o_lO85QksbvE"
      },
      "source": [
        "### BBMS3004 2026 - Targeted sequencing in acute myeloid leukaemia\n",
        "\n",
        "*by Alvin Ip*\n",
        "\n",
        "This practical aims to demonstrate how next-generation sequencing data is processed to yield clinically useful information.\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "7gnJkcA9tTBn"
      },
      "source": [
        "## *** Package installation and downloads for workshop (~ 10 minutes)\n",
        "\n",
        "1.   conda (for simple installation of packages)\n",
        "2.   FastQC (for reads quality check)\n",
        "3.   bwa (tools for sequence alignment)\n",
        "4.   samtools (tools for processing sam & bam files)  \n",
        "\n",
        "**IMPORTANT：Every time you connect to Google Colab, you have to perform these set up steps again.**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "AOrw7MhPztUW"
      },
      "outputs": [],
      "source": [
        "# Set working pathway to your own google drive (~ 1 min)\n",
        "from google.colab import drive\n",
        "drive.mount('/content/gdrive')"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "B_5lJh21K7Ya"
      },
      "outputs": [],
      "source": [
        "# Install igv-notebook (<1 min)\n",
        "!pip install igv-notebook"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "UJMUyjxntquy"
      },
      "outputs": [],
      "source": [
        "# Install conda (~ 1 min). There will be a message saying that the session has crashed, but don't worry about this. This is due to the session restarting following conda installation.\n",
        "!pip install -q condacolab\n",
        "import condacolab\n",
        "condacolab.install()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "F5J8UCIn7i7H"
      },
      "outputs": [],
      "source": [
        "# Install fastqc (~ 2 mins)\n",
        "!conda install -c bioconda fastqc"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "hkiiT0YRz4S2"
      },
      "outputs": [],
      "source": [
        "# Install bwa (~ 1 min)\n",
        "!conda install -c bioconda bwa"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "oMH5G73bz_CI"
      },
      "outputs": [],
      "source": [
        "# Install samtools (~1 min)\n",
        "#!conda install -c bioconda samtools"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Install samtools, bcftools, htslib, bedtools, bamtools (~ 2 mins)\n",
        "!conda install -y -c conda-forge -c bioconda \\\n",
        "  samtools=1.20 bcftools=1.20 htslib=1.20 bedtools=2.31.1 bamtools"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Check installation ran correctly\n",
        "!samtools --version | head -1"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Install GATK in a new environment (to specify the exapct openjdk and python version) ~1.5 mins\n",
        "!conda create -y -n gatkenv python=3.10\n",
        "!conda config --env --set channel_priority strict\n",
        "!conda install -y -n gatkenv -c conda-forge -c bioconda gatk4=4.6.2.0 openjdk=17"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Check that gatk is installed properly\n",
        "!conda run -n gatkenv gatk --version"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Install igv-notebook\n",
        "import sys\n",
        "print(sys.version, sys.executable)\n",
        "!{sys.executable} -m pip install -U igv-notebook\n",
        "import igv_notebook"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "jQ4sstuDsBO1"
      },
      "source": [
        "## Set working directory\n",
        "\n",
        "By default, the working directory will be My Drive/PB_course"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "p4_KlYtbrzzF"
      },
      "outputs": [],
      "source": [
        "import os\n",
        "try:\n",
        "  os.mkdir(\"/content/gdrive/My Drive/PB_course\")         # change this path if necessary\n",
        "except FileExistsError:\n",
        "  print(\"directory already exist. OK to continue\")\n",
        "os.chdir(\"/content/gdrive/My Drive/PB_course\")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "TPs7Tb-tyxc_"
      },
      "source": [
        "## Download ready prepared files for analysis."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "collapsed": true,
        "id": "OwoP1gu3GWLX"
      },
      "outputs": [],
      "source": [
        "# Download reference sequence\n",
        "# Double check that we are in the right directory (~ 30s)\n",
        "import os\n",
        "os.chdir(\"/content/gdrive/MyDrive/PB_course\")                     # change this path if necessary\n",
        "\n",
        "import os\n",
        "if os.path.isfile(\"/content/gdrive/MyDrive/PB_course/DB_trunc/chr2.fa\"):    # check if the file exist\n",
        "  print(\"reference file already exist, OK to continue.\")\n",
        "else:\n",
        "  !pip install gdown\n",
        "  !gdown -O DB_trunc.zip https://drive.google.com/uc?id=1aRJVznjy5WLQ5Dc0DT9c6NiXw64HdoKr # download if file not exist\n",
        "  # unzip fasta file\n",
        "  !unzip -o DB_trunc.zip\n",
        "  # remove the zip file after extraction\n",
        "  !rm DB_trunc.zip\n",
        "\n",
        "!ls -l ./DB_trunc/"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "MsTqxxiP0AzQ"
      },
      "outputs": [],
      "source": [
        "# Download sample sequences\n",
        "import os\n",
        "os.chdir(\"/content/gdrive/My Drive/PB_course/\")\n",
        "if os.path.isfile(\"/content/gdrive/MyDrive/PB_course/Datasets/ChIP-seq_H3K27ac_example.fq.gz\"):    # check if the file exist\n",
        "  print(\"file already exist, OK to continue.\")\n",
        "else:\n",
        " !wget -O Datasets.zip https://github.com/jasonwong-lab/HKU-Practical-Bioinformatics/raw/main/files/Datasets.zip    # download necessary file\n",
        " !unzip -o Datasets.zip   #unzip file\n",
        " !rm Datasets.zip"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "XB9TrCTC7Ncy"
      },
      "source": [
        "## NGS_alignment command line\n",
        "\n",
        "1.1 Quality control\n",
        "\n",
        "1.2. Sequence alignment using Burrows–Wheeler Aligner (BWA)\n",
        "\n",
        "1.3. Viewing SAM files\n",
        "\n",
        "1.4. Align the WGS paired-end file"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "4dYhhxiG4lMM"
      },
      "outputs": [],
      "source": [
        "# Check what datasets we have downloaded\n",
        "%cd /content/gdrive/MyDrive/PB_course\n",
        "!ls -l\n",
        "\n",
        "%cd Datasets/\n",
        "!ls -l"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "EMrO9wOVnNH2"
      },
      "outputs": [],
      "source": [
        "# To look at the ChIP-seq fastq file type (head -n 12 to print first 12 lines):\n",
        "!zcat < ChIP-seq_H3K27ac_example.fq.gz | head -n 12"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "t4g-To8mPw8V"
      },
      "outputs": [],
      "source": [
        "# How many reads are there in the file? Type your code below:\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "pxBspjPp7Dms"
      },
      "source": [
        "### Quality control"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "f82vAykr7Fie"
      },
      "outputs": [],
      "source": [
        "## Run FastQC\n",
        "!fastqc ChIP-seq_H3K27ac_example.fq.gz"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "7v4bH04sZtn5"
      },
      "outputs": [],
      "source": [
        "# Check the o html file\n",
        "!ls\n",
        "\n",
        "# Download the html file and check it on your local browser\n",
        "from google.colab import files\n",
        "files.download('ChIP-seq_H3K27ac_example_fastqc.html')"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Gpt-qEVO81Bs"
      },
      "source": [
        "You can also download the file **ChIP-seq_H3K27ac_example_fastqc.html** from \"gdrive/MyDrive/PB_course/Datasets\" on the left side."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "mwdi1hJ59Dtu"
      },
      "source": [
        "### Burrows–Wheeler Aligner"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "yE-r2cjw8NfS"
      },
      "outputs": [],
      "source": [
        "# Look at the options for bwa and bwa mem\n",
        "!bwa"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "EQCq422B9HbZ"
      },
      "outputs": [],
      "source": [
        "!bwa mem"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "AuteGS6HdjLs"
      },
      "outputs": [],
      "source": [
        "# Let's take a look at the database files\n",
        "%cd /content/gdrive/My Drive/PB_course/DB_trunc\n",
        "\n",
        "!ls -l"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "aFb8h3SZNyjY"
      },
      "outputs": [],
      "source": [
        "# Get ready to run BWA: First go into the Datasets directory\n",
        "%cd /content/gdrive/My Drive/PB_course/Datasets"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "36JkHL9IeUkd"
      },
      "outputs": [],
      "source": [
        "# Make a directory to store the output file\n",
        "!mkdir BAM"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "8EBR94fLMsQI"
      },
      "outputs": [],
      "source": [
        "# Do sequence alignment with the default options\n",
        "!bwa mem ../DB_trunc/chr2.fa ./ChIP-seq_H3K27ac_example.fq.gz > ./BAM/ChIP-seq_H3K27ac_example.sam"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "8oqpLEwbNsCd"
      },
      "outputs": [],
      "source": [
        "# Check the result\n",
        "!head -n 20 ./BAM/ChIP-seq_H3K27ac_example.sam"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "BEQNtRCyOSOk"
      },
      "source": [
        "### Working with SAM files"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "KLBxHt5COTho"
      },
      "outputs": [],
      "source": [
        "# Check the samtools command\n",
        "!samtools"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "6Pgw5h7JOXMO"
      },
      "outputs": [],
      "source": [
        "# Check out some stats about our aligned file\n",
        "!samtools flagstat ./BAM/ChIP-seq_H3K27ac_example.sam"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "A2PFyYk3OjWJ"
      },
      "outputs": [],
      "source": [
        "# Prepare the file for viewing on genome browser:\n",
        "#Step 1 – convert SAM to BAM\n",
        "!samtools view -b ./BAM/ChIP-seq_H3K27ac_example.sam > ./BAM/ChIP-seq_H3K27ac_example.bam"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "IguX9XqQOwf9"
      },
      "outputs": [],
      "source": [
        "# Step 2 – sort BAM file\n",
        "!samtools sort ./BAM/ChIP-seq_H3K27ac_example.bam > ./BAM/ChIP-seq_H3K27ac_example_sorted.bam"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "D92KppCUFX9R"
      },
      "outputs": [],
      "source": [
        "# Step 3 – index BAM file\n",
        "!samtools index ./BAM/ChIP-seq_H3K27ac_example_sorted.bam"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "kmJrDMuPF8Fn"
      },
      "outputs": [],
      "source": [
        "# Look at the files that we have created\n",
        "!ls -l ./BAM/"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Jo8cb29WK2pC"
      },
      "source": [
        "### IGV browser"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "a_tqSvF8TTVp"
      },
      "outputs": [],
      "source": [
        "# Load track from local paths\n",
        "import igv_notebook\n",
        "\n",
        "igv_notebook.init()\n",
        "\n",
        "b = igv_notebook.Browser(\n",
        "    {\n",
        "        \"genome\": \"hg38\",\n",
        "        \"locus\": \"chr2:47,782,081-47,807,953\"\n",
        "    }\n",
        ")\n",
        "\n",
        "b.load_track(\n",
        "    {\n",
        "        \"name\": \"ChIP-seq_H3K27ac\",\n",
        "        \"path\": \"./BAM/ChIP-seq_H3K27ac_example_sorted.bam\",\n",
        "        \"indexPath\": \"./BAM/ChIP-seq_H3K27ac_example_sorted.bam.bai\",\n",
        "        \"format\": \"bam\",\n",
        "        \"type\": \"alignment\"\n",
        "    })"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "nRQZFpflGGhN"
      },
      "source": [
        "### Align the WGS paired-end file"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "jXpX831tnQp9"
      },
      "outputs": [],
      "source": [
        "# Check that we have the WGS fastq files\n",
        "%cd /content/gdrive/My Drive/PB_course/Datasets\n",
        "!ls -l"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "vtckkwSm7Hw8"
      },
      "outputs": [],
      "source": [
        "# If you don't have WGS_example_1.fq.gz or WGS_example_2.fq.gz, run this cell to redownload the files\n",
        "import os\n",
        "os.chdir(\"/content/gdrive/My Drive/PB_course/\")\n",
        "\n",
        "import os\n",
        "if os.path.isfile(\"/content/gdrive/MyDrive/PB_course/Datasets/WGS_example_1.fq.gz\"):    # check if the file exist\n",
        "  print(\"reference file already exist, OK to continue.\")\n",
        "else:\n",
        " !wget -O Datasets.zip https://github.com/jasonwong-lab/HKU-Practical-Bioinformatics/raw/main/files/Datasets.zip\n",
        " !unzip -o Datasets.zip   #unzip file\n",
        " !rm Datasets.zip\n",
        "\n",
        "# Current directory should still be ~/Datasets\n",
        "%cd Datasets/\n",
        "!ls -l"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "l6l-vb9cGILx"
      },
      "outputs": [],
      "source": [
        "# Look at our work directory and paired-end file:\n",
        "!zcat < WGS_example_1.fq.gz |head -1\n",
        "!zcat < WGS_example_2.fq.gz |head -1"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "Eq9VuU3BGl1U"
      },
      "outputs": [],
      "source": [
        "# Align the WGS paired-end file:( ~ 1 mins)\n",
        "# Step 1 – Aligning paired-end file using bwa mem:\n",
        "!bwa mem ../DB_trunc/chr2.fa ./WGS_example_1.fq.gz ./WGS_example_2.fq.gz > ./BAM/WGS_example.sam\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "162y1PgfHAQP"
      },
      "outputs": [],
      "source": [
        "# Step 2 – Output sorted BAM, this time use piping to skip one step:\n",
        "!samtools view -b ./BAM/WGS_example.sam | samtools sort >./BAM/WGS_example_sorted.bam"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "Xbws6jwVHXaQ"
      },
      "outputs": [],
      "source": [
        "# Step 3 – Index sorted bam file\n",
        "!samtools index ./BAM/WGS_example_sorted.bam"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "YjwwnUF2HcQo"
      },
      "outputs": [],
      "source": [
        "# Check stats about the aligned WGS file\n",
        "!samtools flagstat ./BAM/WGS_example_sorted.bam"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "evZ_KRIwca9N"
      },
      "outputs": [],
      "source": [
        "## IGV\n",
        "# Load track from local paths\n",
        "import igv_notebook\n",
        "\n",
        "igv_notebook.init()\n",
        "\n",
        "b = igv_notebook.Browser(\n",
        "    {\n",
        "        \"genome\": \"hg38\",\n",
        "        \"locus\": \"chr2:47,781,678-47,809,690\"\n",
        "    }\n",
        ")\n",
        "\n",
        "b.load_track(\n",
        "    {\n",
        "        \"name\": \"ChIP-seq H3K27ac\",\n",
        "        \"path\": \"./BAM/ChIP-seq_H3K27ac_example_sorted.bam\",\n",
        "        \"indexPath\": \"./BAM/ChIP-seq_H3K27ac_example_sorted.bam.bai\",\n",
        "        \"format\": \"bam\",\n",
        "        \"type\": \"alignment\"\n",
        "    })\n",
        "\n",
        "b.load_track(\n",
        "    {\n",
        "        \"name\": \"WGS\",\n",
        "        \"path\": \"./BAM/WGS_example_sorted.bam\",\n",
        "        \"indexPath\": \"./BAM/WGS_example_sorted.bam.bai\",\n",
        "        \"format\": \"bam\",\n",
        "        \"type\": \"alignment\"\n",
        "    })"
      ]
    }
  ],
  "metadata": {
    "colab": {
      "include_colab_link": true,
      "provenance": []
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}
