{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "colab_type": "text",
        "id": "view-in-github"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/jasonwong-lab/HKU-Practical-Bioinformatics/blob/main/NGS_sequence_alignment_command_line.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "o_lO85QksbvE"
      },
      "source": [
        "## BBMS3004 - Targeted sequencing in acute myeloid leukaemia\n",
        "\n",
        "*6 February 2026*\n",
        "\n",
        "*Alvin Ip*\n",
        "\n",
        "This practical aims to demonstrate how next-generation sequencing data is processed to yield clinically useful information. It is a highly condensed version of BIOF3002 Clinical Genomics and Bioinformatics that Prof Jason Wong and I co-coordinated for bioinformatics students.\n",
        "\n",
        "In this session. we will build a toy bioinformatic pipeline to get a feel of what it is like to practice bioinformatics using Linux. Our platform will be [Google Colab](https://colab.research.google.com/), which is a free cloud computing platform. It uses Jupyter notebook, a convenient platform for programming in Python. As we will be using Linux commands in our session, we will exploit the \"!\" command in Python function to turn our script into system command expressed in Linux. Don't worry. This session does not assume any programming skills from you. The Jupyter notebook setup will take you through the analysis. It would be very beneficial if you can slowly read through the comments (i.e. those lines starting with \"#\") and the commands (i.e. the lines that do not start with \"#\") before you run them and try to understand what they are doing.\n",
        "\n",
        "### GATK Best Practice for Somatic SNVs and Indels\n",
        "\n",
        "We will try to build a pipeline according [GATK best practices for data preprocessing](https://gatk.broadinstitute.org/hc/en-us/articles/360035535912) and [GATK best practices for somatic SNVs and indels](https://gatk.broadinstitute.org/hc/en-us/articles/360035894731-Somatic-short-variant-discovery-SNVs-Indels-). Other GATK Best Practices pipelines can be found in the menu of the website as well.\n",
        "\n",
        "### Building the Toy Pipeline\n",
        "\n",
        "While we are trying to build a functional bioinformatic pipeline, but as some elements of this pipeline has been abridged to suit the computer resources and running time in the teaching session, this pipeline should NOT be used for any other purposes, including clinical practice and research."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "7gnJkcA9tTBn"
      },
      "source": [
        "### Package installation and downloads for workshop (~ 5-10 minutes)\n",
        "\n",
        "1.   conda (for simple installation of packages)\n",
        "2.   FastQC (for reads quality check)\n",
        "3.   Trimmomatic (for trimming low-quality bases and adapters)\n",
        "4.   bwa (tools for sequence alignment)\n",
        "5.   samtools (tools for processing sam & bam files)\n",
        "6.   IGV (for visualising sequencing data and variants)  \n",
        "\n",
        "**IMPORTANT：Every time you connect to Google Colab, you have to perform these set up steps again.**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Install terminal for Google Colab (optional)\n",
        "#!pip install colab-xterm\n",
        "#%load_ext colabxterm\n",
        "#%xterm"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "AOrw7MhPztUW"
      },
      "outputs": [],
      "source": [
        "# Set working pathway to your own Google drive, so that the data files are saved after closing Colab (~ 1 min)\n",
        "from google.colab import drive\n",
        "drive.mount('/content/gdrive')"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "B_5lJh21K7Ya"
      },
      "outputs": [],
      "source": [
        "# Install igv-notebook (<1 min)\n",
        "!pip install igv-notebook"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "UJMUyjxntquy"
      },
      "outputs": [],
      "source": [
        "# Install conda (~ 1 min). There will be a message saying that the session has crashed, but don't worry about this. This is due to the session restarting following conda installation.\n",
        "!pip install -q condacolab\n",
        "import condacolab\n",
        "condacolab.install()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "F5J8UCIn7i7H"
      },
      "outputs": [],
      "source": [
        "# Install fastqc (~ 2 mins)\n",
        "!conda install -c bioconda fastqc"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "hkiiT0YRz4S2"
      },
      "outputs": [],
      "source": [
        "# Install bwa (~ 1 min)\n",
        "!conda install -c bioconda bwa"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Install samtools, bcftools, htslib, bedtools, bamtools (~ 2 mins)\n",
        "!conda install -y -c conda-forge -c bioconda \\\n",
        "  samtools=1.20 bcftools=1.20 htslib=1.20 bedtools=2.31.1 bamtools"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Check installation ran correctly\n",
        "!samtools --version | head -1"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Install GATK in a new environment (to specify the exapct openjdk and python version) ~1.5 mins\n",
        "!conda create -y -n gatkenv python=3.10\n",
        "!conda config --env --set channel_priority strict\n",
        "!conda install -y -n gatkenv -c conda-forge -c bioconda gatk4=4.6.2.0 openjdk=17"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Check that gatk is installed properly\n",
        "!conda run -n gatkenv gatk --version"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "jQ4sstuDsBO1"
      },
      "source": [
        "### Set working directory\n",
        "\n",
        "By default, the working directory will be My Drive/bbms3004"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "p4_KlYtbrzzF"
      },
      "outputs": [],
      "source": [
        "import os\n",
        "try:\n",
        "  os.mkdir(\"/content/gdrive/My Drive/bbms3004\")         # change this path if necessary\n",
        "except FileExistsError:\n",
        "  print(\"directory already exist. OK to continue\")\n",
        "os.chdir(\"/content/gdrive/My Drive/bbms3004\")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "TPs7Tb-tyxc_"
      },
      "source": [
        "### Download the full toolkit that includes reference files, some tools and sequencing data for analysis."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "collapsed": true,
        "id": "OwoP1gu3GWLX"
      },
      "outputs": [],
      "source": [
        "# Download data files from Google Drive if not already present (~ 3 min)\n",
        "# Double check that we are in the right directory (~ 30s)\n",
        "import os\n",
        "os.chdir(\"/content/gdrive/MyDrive/bbms3004\")                     # change this path if necessary\n",
        "\n",
        "import os\n",
        "if os.path.isfile(\"/content/gdrive/MyDrive/bbms3004/bbms3004_toolkit/17H0510082_1.fastq.gz\"):    # check if the file exist\n",
        "  print(\"Data files already exist, OK to continue.\")\n",
        "else:\n",
        "  !pip install gdown\n",
        "  !gdown -O bbms3004_toolkit.zip https://drive.google.com/uc?id=1TzOHbEKF2P1EWoxVsS_pvzS4ZS46XNXh # download if file not exist\n",
        "\n",
        "  # unzip toolkit files\n",
        "  !unzip -o bbms3004_toolkit.zip\n",
        "\n",
        "  # remove the downloaded zip file after extraction to save space\n",
        "  !rm bbms3004_toolkit.zip\n",
        "\n",
        "# Check that the data files are present\n",
        "!ls -l ./bbms3004_toolkit/"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "XB9TrCTC7Ncy"
      },
      "source": [
        "## Start processing the NGS sequencing data following these steps:\n",
        "\n",
        "1. Quality control\n",
        "\n",
        "2. Trimming of adapters and low-quality bases\n",
        "\n",
        "3. Sequence alignment using Burrows–Wheeler Aligner (BWA)\n",
        "\n",
        "4. Viewing SAM files\n",
        "\n",
        "5. Mark and remove PCR duplicates\n",
        "\n",
        "6. Base recalibration\n",
        "\n",
        "7. Variant calling by Mutect2\n",
        "\n",
        "8. Variant filtering by bcftools\n",
        "\n",
        "9. Variant annotation by ANNOVAR\n",
        "\n",
        "10. Manual curation of variants\n",
        "\n",
        "11. Visualisation of curated variants in IGV\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "4dYhhxiG4lMM"
      },
      "outputs": [],
      "source": [
        "# Check what datasets we have downloaded\n",
        "%cd /content/gdrive/MyDrive/bbms3004/bbms3004_toolkit\n",
        "!ls -l\n",
        "\n",
        "%cd data\n",
        "!ls -l"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "EMrO9wOVnNH2"
      },
      "outputs": [],
      "source": [
        "# To look at the ChIP-seq fastq file type (head -n 12 to print first 12 lines):\n",
        "!zcat 17H0510082_1.fastq.gz | head -n 12"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "t4g-To8mPw8V"
      },
      "outputs": [],
      "source": [
        "# How many reads are there in the file? Type your code below:\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "pxBspjPp7Dms"
      },
      "source": [
        "### 1. Quality control"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "f82vAykr7Fie"
      },
      "outputs": [],
      "source": [
        "## Run FastQC\n",
        "!fastqc 17H0510082_1.fastq.gz\n",
        "!fastqc 17H0510082_2.fastq.gz"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "7v4bH04sZtn5"
      },
      "outputs": [],
      "source": [
        "# Check the html file output by FastQC\n",
        "!ls\n",
        "\n",
        "# Download the html file and check it on your local browser\n",
        "from google.colab import files\n",
        "files.download('17H0510082_1_fastqc.html')"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Gpt-qEVO81Bs"
      },
      "source": [
        "You can also download the file **17H0510082_1_fastqc.html** from \"gdrive/MyDrive/bbms3004/Datasets\" on the left side."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "### 2. Trimming of adapters and low-quality bases"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Perform Trimmomatic\n",
        "!java -jar ../Trimmomatic-0.39/trimmomatic-0.39.jar PE -phred33 \\\n",
        "    17H0510082_1.fastq.gz 17H0510082_2.fastq.gz \\\n",
        "    17H0510082_1_trimmed_paired.fq.gz 17H0510082_1_trimmed_unpaired.fq.gz 17H0510082_2_trimmed_paired.fq.gz 17H0510082_2_trimmed_unpaired.fq.gz \\\n",
        "    ILLUMINACLIP:../Trimmomatic-0.39/adapters/TruSeq3-PE-2.fa:2:30:10 LEADING:10 TRAILING:10 SLIDINGWINDOW:4:15 MINLEN:40\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "mwdi1hJ59Dtu"
      },
      "source": [
        "### 3. Sequence alignment using Burrows–Wheeler Aligner (BWA)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "yE-r2cjw8NfS"
      },
      "outputs": [],
      "source": [
        "# Look at the options for bwa and bwa mem\n",
        "!bwa"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "EQCq422B9HbZ"
      },
      "outputs": [],
      "source": [
        "!bwa mem"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "AuteGS6HdjLs"
      },
      "outputs": [],
      "source": [
        "# Let's take a look at the database files\n",
        "%cd /content/gdrive/My Drive/bbms3004/bbms3004_toolkit/Ref\n",
        "\n",
        "!ls -l"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "aFb8h3SZNyjY"
      },
      "outputs": [],
      "source": [
        "# Get ready to run BWA: First go into the Datasets directory\n",
        "%cd /content/gdrive/My Drive/bbms3004/bbms3004_toolkit/data"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "8EBR94fLMsQI"
      },
      "outputs": [],
      "source": [
        "# Do sequence alignment with the default options\n",
        "!bwa mem -t 4 -M ../Ref/ucsc.hg19.subset.fasta 17H0510082_1_trimmed_paired.fq.gz 17H0510082_2_trimmed_paired.fq.gz > 17H0510082.sam"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "### 4. Viewing SAM files"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "8oqpLEwbNsCd"
      },
      "outputs": [],
      "source": [
        "# Check the result\n",
        "!head -n 20 17H0510082.sam"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Sort the SAM file and compress SAM to BAM\n",
        "!conda run -n gatkenv gatk SortSam \\\n",
        "    -I 17H0510082.sam \\\n",
        "    -O 17H0510082_sorted.bam \\\n",
        "    --SORT_ORDER coordinate"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Add read groups (RG), tags for downstream tools to run properly\n",
        "!conda run -n gatkenv gatk AddOrReplaceReadGroups \\\n",
        "    -I 17H0510082_sorted.bam \\\n",
        "    -O 17H0510082_RG.bam \\\n",
        "    --RGID SPACE --RGLB panel --RGPL ILLUMINA --RGPU unit1 --RGSM 17H0510082 --CREATE_INDEX true"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "### 5. Mark and remove PCR duplicates"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Perform MarkDuplicates to mark duplicate reads\n",
        "!conda run -n gatkenv gatk MarkDuplicates \\\n",
        "    -I 17H0510082_RG.bam \\\n",
        "    -O 17H0510082_MD.bam \\\n",
        "    --METRICS_FILE 17H0510082_MD.stats \\\n",
        "    --CREATE_INDEX true"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Inspect the duplicate rate of the sequencing data\n",
        "# Try to locate the column with header \"PERCENT_DUPLICATION\"\n",
        "!cat 17H0510082_MD.stats | sed -n '7,8p' | column -t -s $'\\t'   # Extract the 7th and 8th lines and format as table"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "### 6. Base recalibration"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Perform BaseRecalibrator to recalibrate base quality scores\n",
        "!conda run -n gatkenv gatk BaseRecalibrator \\\n",
        "    -I 17H0510082_MD.bam \\\n",
        "    -R ../Ref/ucsc.hg19.subset.fasta \\\n",
        "    --known-sites ../Ref/dbsnp_138.hg19.vcf \\\n",
        "    --known-sites ../Ref/Mills_and_1000G_gold_standard.indels.hg19.vcf \\\n",
        "    -L ../Ref/myeloid-targets.interval_list \\\n",
        "    -ip 50 \\\n",
        "    -O 17H0510082_recal_data.table\n",
        "\n",
        "# Apply the recalibration to the sequence data\n",
        "!conda run -n gatkenv gatk ApplyBQSR \\\n",
        "    -R ../Ref/ucsc.hg19.subset.fasta \\\n",
        "    -I 17H0510082_MD.bam \\\n",
        "    --bqsr-recal-file 17H0510082_recal_data.table \\\n",
        "    -O 17H0510082_BR.bam"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "#### Collect quality metrics before variant calling"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Collect quality control metrics using the variant-call ready BAM file (~ 3 minutes)\n",
        "# Collect QC metrics\n",
        "!conda run -n gatkenv gatk CollectMultipleMetrics \\\n",
        "    -I 17H0510082_BR.bam \\\n",
        "    -R ../Ref/ucsc.hg19.subset.fasta \\\n",
        "    -O 17H0510082_QC_metrics\n",
        "\n",
        "# Collect read counts\n",
        "!conda run -n gatkenv gatk CollectReadCounts \\\n",
        "    -I 17H0510082_BR.bam \\\n",
        "    -L ../Ref/myeloid-targets.interval_list \\\n",
        "    --interval-merging-rule OVERLAPPING_ONLY \\\n",
        "    --format TSV \\\n",
        "    -O 17H0510082_read_counts.table\n",
        "\n",
        "# Collect HS metrics\n",
        "!conda run -n gatkenv gatk CollectHsMetrics \\\n",
        "    -I 17H0510082_BR.bam \\\n",
        "    -O 17H0510082_hs_metrics.txt \\\n",
        "    -R ../Ref/ucsc.hg19.subset.fasta \\\n",
        "    -BAIT_INTERVALS ../Ref/myeloid-probe-coords.interval_list \\\n",
        "    -TARGET_INTERVALS ../Ref/myeloid-targets.interval_list"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Inspect the output from HS metrics\n",
        "!cat 17H0510082_hs_metrics.txt | sed -n '7,8p' | awk '{print $13 \"\\t\" $23 \"\\t\" $34 \"\\t\" $45 \"\\t\" $48 \"\\t\" $52 \"\\t\" $53 \"\\t\" $55 \"\\t\" $56 \"\\t\" $58}' | column -t -s $'\\t'"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "### 7. Variant calling by Mutect2"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Perform variant calling with Mutect2 (~ 5 minutes)\n",
        "!conda run -n gatkenv gatk Mutect2 \\\n",
        "    -R ../Ref/ucsc.hg19.subset.fasta \\\n",
        "    -I 17H0510082_BR.bam \\\n",
        "    -L ../Ref/myeloid-targets.interval_list \\\n",
        "    --germline-resource ../Ref/af-only-gnomad.myeloid.bedtools.vcf.gz \\\n",
        "    --f1r2-tar-gz f1r2.tar.gz \\\n",
        "    -O 17H0510082_unfiltered.vcf.gz\n",
        "\n",
        "!conda run -n gatkenv gatk LearnReadOrientationModel \\\n",
        "    -I f1r2.tar.gz \\\n",
        "    -O read-orientation-model.tar.gz\n",
        "\n",
        "!conda run -n gatkenv gatk GetPileupSummaries \\\n",
        "    -I 17H0510082_BR.bam \\\n",
        "    -V ../Ref/small_exac_common_myeloid.vcf.gz \\\n",
        "    -L ../Ref/small_exac_common_myeloid.vcf.gz  \\\n",
        "    -O getpileupsummaries.table\n",
        "\n",
        "!conda run -n gatkenv gatk CalculateContamination \\\n",
        "    -I getpileupsummaries.table \\\n",
        "    -O contamination.table\n",
        "\n",
        "!conda run -n gatkenv gatk FilterMutectCalls \\\n",
        "    -V 17H0510082_unfiltered.vcf.gz \\\n",
        "    -R ../Ref/ucsc.hg19.subset.fasta \\\n",
        "    --contamination-table contamination.table \\\n",
        "    --ob-priors read-orientation-model.tar.gz \\\n",
        "    -O 17H0510082_filtered.vcf.gz"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Inspect the VCF file from Mutect2\n",
        "!zcat 17H0510082_filtered.vcf.gz"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "### 8. Variant filtering by bcftools"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Use bcftools to filter variants with variant allele frequencies less than 0.05\n",
        "!bcftools filter -i \" FORMAT/AF > 0.05 \" 17H0510082_filtered.vcf.gz -o 17H0510082_filtered_AF05.vcf"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "### 9. Variant annotation by ANNOVAR"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Perform variant annotation by ANNOVAR\n",
        "!chmod -R 700 ../annovar   # give execute permission to annovar scripts\n",
        "!perl ../annovar/table_annovar.pl 17H0510082_filtered_AF05.vcf \\\n",
        "    ../annovar/humandb/ \\\n",
        "    -buildver hg19 \\\n",
        "    -out 17H0510082_filtered_annotate \\\n",
        "    -remove \\\n",
        "    -protocol refGene,cosmic86,clinvar_20170905,exac03nontcga,gnomad_exome \\\n",
        "    -operation g,f,f,f,f \\\n",
        "    -nastring . \\\n",
        "    -vcfinput"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "### 10. Manual curation of variants"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Download ANNOVAR annotated file for viewing and curation\n",
        "from google.colab import files\n",
        "files.download('17H0510082_filtered_annotate.hg19_multianno.txt')"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "### 11. Visualisation of curated variants in IGV"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "## IGV\n",
        "# Load track from local paths\n",
        "import igv_notebook\n",
        "\n",
        "igv_notebook.init()\n",
        "\n",
        "# FLT3\n",
        "b = igv_notebook.Browser(\n",
        "    {\n",
        "        \"genome\": \"hg19\",\n",
        "        \"locus\": \"chr13:28592610-28592670\"\n",
        "    }\n",
        ")\n",
        "\n",
        "b.load_track(\n",
        "    {\n",
        "        \"name\": \"17H0510082\",\n",
        "        \"path\": \"17H0510082_BR.bam\",\n",
        "        \"indexPath\": \"17H0510082_BR.bai\",\n",
        "        \"format\": \"bam\",\n",
        "        \"type\": \"alignment\"\n",
        "    })\n",
        "\n",
        "# NPM1\n",
        "c = igv_notebook.Browser(\n",
        "    {\n",
        "        \"genome\": \"hg19\",\n",
        "        \"locus\": \"chr5:170837516-170837576\"\n",
        "    }\n",
        ")\n",
        "\n",
        "c.load_track(\n",
        "    {\n",
        "        \"name\": \"17H0510082\",\n",
        "        \"path\": \"17H0510082_BR.bam\",\n",
        "        \"indexPath\": \"17H0510082_BR.bai\",\n",
        "        \"format\": \"bam\",\n",
        "        \"type\": \"alignment\"\n",
        "    })\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "### After the session\n",
        "\n",
        "Thank you for going through this session. If you have any inquiry regarding the course content or issues with running the server, please do not hesitate to email me at iphowan@hku.hk"
      ]
    }
  ],
  "metadata": {
    "colab": {
      "include_colab_link": true,
      "provenance": []
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}
